test/e2e/rules_api_test.go

```diff
@@ -33,42 +35,42 @@ func TestRulesAPI_Fanout(t *testing.T) {
 	testutil.Ok(t, err)
 	t.Cleanup(e2ethanos.CleanScenario(t, s))
 
-	rulesSubDir := filepath.Join("rules")
-	testutil.Ok(t, os.MkdirAll(filepath.Join(s.SharedDir(), rulesSubDir), os.ModePerm))
-	createRuleFiles(t, filepath.Join(s.SharedDir(), rulesSubDir))
+	promRulesSubDir := filepath.Join("rules")
+	testutil.Ok(t, os.MkdirAll(filepath.Join(s.SharedDir(), promRulesSubDir), os.ModePerm))
+	// Create the abort_on_partial_response alert for Prometheus.
+	// We don't create the warn_on_partial_response alert as Prometheus has strict yaml unmarshalling.
+	createRuleFile(t, filepath.Join(s.SharedDir(), promRulesSubDir, "rules.yaml"), testAlertRuleAbortOnPartialResponse)
+
+	thanosRulesSubDir := filepath.Join("thanos-rules")
+	testutil.Ok(t, os.MkdirAll(filepath.Join(s.SharedDir(), thanosRulesSubDir), os.ModePerm))
+	createRuleFiles(t, filepath.Join(s.SharedDir(), thanosRulesSubDir))
 
 	// 2x Prometheus.
 	prom1, sidecar1, err := e2ethanos.NewPrometheusWithSidecar(
 		s.SharedDir(),
 		netName,
 		"prom1",
-		defaultPromConfig("ha", 0, "", filepath.Join(e2e.ContainerSharedDir, rulesSubDir, "*.yaml")),
+		defaultPromConfig("ha", 0, "", filepath.Join(e2e.ContainerSharedDir, promRulesSubDir, "*.yaml")),
 		e2ethanos.DefaultPrometheusImage(),
 	)
 	testutil.Ok(t, err)
 	prom2, sidecar2, err := e2ethanos.NewPrometheusWithSidecar(
 		s.SharedDir(),
 		netName,
 		"prom2",
-		defaultPromConfig("ha", 1, "", filepath.Join(e2e.ContainerSharedDir, rulesSubDir, "*.yaml")),
+		defaultPromConfig("ha", 1, "", filepath.Join(e2e.ContainerSharedDir, promRulesSubDir, "*.yaml")),
 		e2ethanos.DefaultPrometheusImage(),
 	)
 	testutil.Ok(t, err)
 	testutil.Ok(t, s.StartAndWaitReady(prom1, sidecar1, prom2, sidecar2))
 
-	// 2x Rulers.
-	r1, err := e2ethanos.NewRuler(s.SharedDir(), "rule1", rulesSubDir, nil, nil)
-	testutil.Ok(t, err)
-	r2, err := e2ethanos.NewRuler(s.SharedDir(), "rule2", rulesSubDir, nil, nil)
-	testutil.Ok(t, err)
-	testutil.Ok(t, s.StartAndWaitReady(r1, r2))
-
 	q, err := e2ethanos.NewQuerier(
 		s.SharedDir(),
 		"query",
-		[]string{sidecar1.GRPCNetworkEndpoint(), sidecar2.GRPCNetworkEndpoint(), r1.GRPCNetworkEndpoint(), r2.GRPCNetworkEndpoint()},
+		// TODO(yeya24): stop hardcoding the GRPC endpoints for rulers.
+		[]string{sidecar1.GRPCNetworkEndpoint(), sidecar2.GRPCNetworkEndpoint(), "rule-rule1:9091", "rule-rule2:9091"},
```

Comment:
Why can't we use endpoint method?

Reply:
If we use endpoints methods of rulers, we need to create and start them first. But to create rulers, we need the address of thanos query first. So either way, we need to hardcode the address for one component.

Reply:
Any comment about this?

Reply:
Well, I wonder if we can improve the framework for this approach. It sounds quite common. 🤔 Right now we put TODO but there is literally no idea how to solve it 🙈

Reply:
Fixed. PTAL